sys/exec.c

```diff
@@ -1,21 +1,162 @@
         if(i % stride == 7) kprintf("\n");
     }
+
+    const Elf32_Ehdr* eh = (Elf32_Ehdr*)elf_image;
+
+    // Start by determining the validity of the elf file.
+
+    // First, check for the magic header.
+    if(eh->e_ident[EI_MAG0] != ELFMAG0 ||
+       eh->e_ident[EI_MAG1] != ELFMAG1 ||
+       eh->e_ident[EI_MAG2] != ELFMAG2 ||
+       eh->e_ident[EI_MAG3] != ELFMAG3){
+        kprintf("[exec] Exec failed: Incorrect ELF magic number\n");
+        return -1;
+    }
+    // Check ELF class
+    if(eh->e_ident[EI_CLASS] != ELFCLASS32){
+        kprintf("[exec] Exec failed: Unsupported ELF class (!= ELF32)\n");
+        return -1;
+    }
+    // Check data format endianess
+    if(eh->e_ident[EI_DATA] != ELFDATA2LSB){
+        kprintf("[exec] Exec failed: ELF file is not low-endian\n");
+        return -1;
+    }
+    // Ignore version and os abi field
+    // Check file type
+    if(eh->e_type != ET_EXEC){
+        kprintf("[exec] Exec failed: ELF is not an executable\n");
+        return -1;
+    }
+    // Check machine architecture field
+    if(eh->e_machine != EM_MIPS){
+        kprintf("[exec] Exec failed: ELF target architecture is not MIPS\n");
+        return -1;
+    }
+
+    // Take note of the entry point
+    kprintf("[exec] Entry point will be at 0x%08x.\n",
+            (unsigned int)eh->e_entry);
+
+    // Ensure prog header size compliance
+    if(eh->e_phentsize != sizeof(Elf32_Phdr)/sizeof(uint8_t)){
+        kprintf("[exec] Exec failed: ELF uses non-standard program header size\n");
+        return -1;
+    }
+
+    // TODO: Take care of assigning thread/process ASID
+    vm_map_t* vmap = vm_map_new(PMAP_USER, 123);
+    // Note: we do not claim ownership of the map
+    set_active_vm_map(vmap);
+
+    // Iterate over prog headers
+    kprintf("[exec] ELF has %d program headers\n", eh->e_phnum);
+
+    const Elf32_Phdr* phs = (Elf32_Phdr*)(elf_image + eh->e_phoff);
+    for(uint8_t i = 0; i < eh->e_phnum; i++){
+        const Elf32_Phdr* ph = phs + i;
+        switch(ph->p_type){
+        case PT_NULL:
+        case PT_NOTE:
+        case PT_PHDR:
+        default:
+            // Ignore the section.
+            break;
+        case PT_DYNAMIC:
+        case PT_INTERP:
+            kprintf("[exec] Exec failed: ELF file requests dynamic linking"
+                    "by providing a PT_DYNAMIC and/or PT_INTERP segment.\n");
+            return -1;
+        case PT_SHLIB:
+            kprintf("[exec] Exec failed: ELF file contains a PT_SHLIB segment\n");
+            return -1;
+        case PT_LOAD:
+            kprintf("[exec] Processing a PT_LOAD segment: VirtAddr = %p, "
+                    "Offset = 0x%08x, FileSiz = 0x%08x, MemSiz = 0x%08x, "
+                    "Flags = %d\n", (void*)ph->p_vaddr, (unsigned int)ph->p_offset,
+                    (unsigned int)ph->p_filesz, (unsigned int)ph->p_memsz,
+                    (unsigned int)ph->p_flags);
+            if(ph->p_vaddr % PAGESIZE){
```

Comment:
FreeBSD checks `p_align`, not `p_vaddr`. `p_align` is more or less page size assumed by the ELF file. Specifically, it is required that `p_vaddr ≡ p_offset (mod p_align)`. This, obviously, doesn't mean that `p_offset` or `p_vaddr` need to be aligned in any way.

I don't see how unaligned `p_vaddr` would be a significant problem. Since the offset inside the page is the same for `p_vaddr` and `p_offset` the segment loading code would just have to round the base address down to the page boundary. True, more data will be loaded from the disk than absolutely necessary, but it is not really a problem.

BTW With PIE being more and more common I wouldn't be so sure that fixed-position ELF binaries are the common case.